### PR TITLE
Origin/departments/controller

### PR DESCRIPTION
**Pull request by amitm144**
made

- linkDeparmentBYId
- getUserByDepartmentId
- deparmentIdBoundry 
- department entity -getters and setters

### CODE REVIEW

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/NewDepartmentBoundary.java
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/NewDepartmentBoundary.java:1]: The class is redundant.
it extends DepartmentBoundary for no reason and it is basically DepartmentIdBoundary.

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/UserBoundary.java
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/UserBoundary.java:1]: Why?

File: usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveDepartmentsController.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    private DepartmentsService departmentsService;
+
+    public ReactiveDepartmentsController(DepartmentsService departmentsService) {
+        this.departmentsService = departmentsService;
+    }
+
+    @PostMapping(
+            consumes = {MediaType.APPLICATION_JSON_VALUE},
+            produces = {MediaType.APPLICATION_JSON_VALUE})
+    public Mono<DepartmentBoundary> createDepartment(@RequestBody NewDepartmentBoundary department) {
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveDepartmentsController.java]:
```suggestion
    public Mono<DepartmentBoundary> createDepartment(@RequestBody DepartmentIdBoundary department) {
```

File: usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package il.ac.afeka.usersservice.data;
+
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+@Document(collection = "departments")
+
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java]:
```suggestion
```

File: usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveUsersController.java [hunk 3]
```diff
@@ -1,64 +1,77 @@
...
 import il.ac.afeka.usersservice.logic.UsersService;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 import org.springframework.web.bind.annotation.*;
+
 import org.springframework.web.server.ResponseStatusException;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
+import java.util.Map;
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveUsersController.java]:
```suggestion
```

Comment (Noam-Levy) [reply usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveUsersController.java]: Remove unused import

File: usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java (hunk 2 above)
Comment (Noam-Levy) [reply usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java]: Delete empty line

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveDepartmentsService.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        this.departmentCrud = departmentCrud;
+    }
+
+    @Override
+    public Mono<DepartmentBoundary> createDepartment(NewDepartmentBoundary department) {
+        return Mono.just(department)
+                .map(NewDepartmentBoundary::toEntity)
+                .flatMap(this.departmentCrud::save)
+                .map(DepartmentBoundary::new);
+    }
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveDepartmentsService.java]: I think that the id should be set by the database and not externally.
You can check how we did it in [MeetApp](https://github.com/Noam-Levy/MeetApp/blob/master/src/main/java/superapp/logic/concreteServices/SuperAppObjectService.java).

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java [hunk 5]
```diff
@@ -1,22 +1,19 @@
...
+public interface UsersService {
+	public Flux<UserBoundary> getUsersByDomain(String domain);
+	public Flux<UserBoundary> getUsersByMinimumAge(int minimumAgeInYears);
+	public Flux<UserBoundary> getUsersByLastname(String lastName);
+	public Flux<UserBoundary> getUserByDepartmentId(String departmentId);
+	public Mono<Void> deleteAll();
+	public Flux<UserBoundary>  getAll();
+	public Mono<UserBoundary> GetUserByEmail(String email, String password);
+	public Mono<UserBoundary> createUser(NewUserBoundary user);
+	public Mono<UserBoundary> linkUserToDepartment(String email, DepartmentIdBoundary departmentId);
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java]:
```suggestion
	public Flux<UserBoundary> getUsersByDomain(String domain);
	
	public Flux<UserBoundary> getUsersByMinimumAge(int minimumAgeInYears);
	
	public Flux<UserBoundary> getUsersByLastname(String lastName);
	
	public Flux<UserBoundary> getUserByDepartmentId(String departmentId);
	
	public Mono<Void> deleteAll();
	
	public Flux<UserBoundary>  getAll();
	
	public Mono<UserBoundary> GetUserByEmail(String email, String password);
	
	public Mono<UserBoundary> createUser(NewUserBoundary user);
	
	public Mono<UserBoundary> linkUserToDepartment(String email, DepartmentIdBoundary departmentId);
	
```

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java [hunk 6]
```diff
@@ -1,22 +1,19 @@
...
 
-public interface UsersService {
-    public Flux<UserBoundary> getUsersByDomain(String domain);
-
-    public Flux<UserBoundary> getUsersByMinimumAge(int minimumAgeInYears);
-
-    public Flux<UserBoundary> getUsersByLastname(String lastName);
-
-    public Mono<Void> deleteAll();
+import java.util.Map;
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java]:
```suggestion
```
Remove unused Import

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveUserCrud.java [hunk 7]
```diff
@@ -13,4 +13,6 @@ public interface ReactiveUserCrud extends ReactiveMongoRepository<UserEntity, St
 	Flux<UserEntity> findAllByLastname( @Param("lastname") String lastname );
 
 	Mono<UserEntity> findByEmailAndPassword( @Param("email") String email, @Param("password") String password );
+
+	Mono<UserEntity> findByEmail( @Param("email") String email);
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveUserCrud.java]: Email is the user Id, you can call findById(email) so this is redundant.

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveUsersService.java [hunk 8]
```diff
@@ -62,4 +76,22 @@ public Mono<UserBoundary> createUser(NewUserBoundary user) {
...
+        return userMono.flatMap(user ->
+                departmentMono.flatMap(departmentEntity -> {
+                            Set<String> rolesSet = new HashSet<>(Arrays.asList(user.getRoles())); // make it set to neglect duplicate
+                            rolesSet.add(departmentEntity.getDeptId()); //add new depID
+                            String[] updatedRoles = rolesSet.toArray(new String[0]); //make it string array
+                            user.setRoles(updatedRoles);
+                            return this.userCrud.save(user);
+                        })
+                        .map(savedUser -> new UserBoundary(savedUser))
+        );
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveUsersService.java:94]: a more elegant way to do that is to set the roles as a set in the user entity and than changing only the supporting functions such as toEntity.

**Review by Noam-Levy — CHANGES_REQUESTED** [review]
All in all looks good. Haven't checked against swagger to see the functionality yet.
will do that as part of my error handling task.

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/UserBoundary.java
Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/boundaries/UserBoundary.java:1]: why what ?

Comment (Noam-Levy) [reply usersService/src/main/java/il/ac/afeka/usersservice/boundaries/UserBoundary.java:1]: Oops I thought it's going to show the file.
these changes do nothing.. why are they here?

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentIdBoundary.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package il.ac.afeka.usersservice.boundaries;
+
+public class DepartmentIdBoundary {
+
+	private String depId ;
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentIdBoundary.java]:
```suggestion
	private String deptId ;
```

Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentIdBoundary.java]: done

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveUserCrud.java (hunk 7 above)
Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveUserCrud.java]: done

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java (hunk 6 above)
Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java]: done

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java (hunk 5 above)
Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/logic/UsersService.java]: done

File: usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveUsersController.java (hunk 3 above)
Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/controllers/ReactiveUsersController.java]: done

File: usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java (hunk 2 above)
Comment (amitm144) [reply usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java]: done

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentIdBoundary.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+		this.deptId = deptId;
+	}
+
+	@Override
+	public String toString() {
+		return "DepartmentIdBoundary{" +
+				"depId='" + deptId + '\'' +
+				'}';
+	}
+}
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentIdBoundary.java:27]: The fix is good, but the boundary is still lacking. 
look at the differences between what's the body structure in swagger and what Eyal have requested:
image

File: usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    public void setDepartmentName(String departmentName) {
+        this.departmentName = departmentName;
+    }
+
+    public String getCreationDate() {
+        return creationDate;
+    }
+
+    public void setCreationDate(String creationDate) {
+        this.creationDate = creationDate;
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/data/DepartmentEntity.java:37]: creation date should be determined by the system and not by the user.

**Review by Noam-Levy — CHANGES_REQUESTED** [review]

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentBoundary.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+    public void setCreationDate(String creationDate) {
+        this.creationDate = creationDate;
+    }
+
+    public DepartmentEntity toEntity() {
+        DepartmentEntity entity = new DepartmentEntity();
+        entity.setDeptId(this.getDeptId());
+        entity.setDepartmentName(this.getDepartmentName());
+        entity.setCreationDate(this.getCreationDate());
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentBoundary.java]:
```suggestion
        entity.setCreationDate(LocalDate.parse("dd-MM-yyyy").toString());
```
Creation date should be determined by the system and not by the user.

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentBoundary.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    public static String changeDateFormat(String inputDate) {
+        // Parse the input date string to a LocalDate object
+        LocalDate localDate = LocalDate.parse(inputDate, DateTimeFormatter.ISO_DATE);
+
+        // Format the LocalDate object to the desired output format
+        String outputDate = localDate.format(DateTimeFormatter.ofPattern("dd-MM-yyyy"));
+
+        return outputDate;
+    }
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentBoundary.java]: This is the wrong place for a utility function such as this one.
This should be in DateValidator and I would also change its name for something more appropriate

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentWrapperBoundary.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
+package il.ac.afeka.usersservice.boundaries;
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentWrapperBoundary.java:1]: Looks good. well done.

File: usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveDepartmentCrud.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package il.ac.afeka.usersservice.logic;
+
+import il.ac.afeka.usersservice.data.DepartmentEntity;
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository;
+import org.springframework.data.repository.query.Param;
+import reactor.core.publisher.Mono;
+
+public interface ReactiveDepartmentCrud extends ReactiveMongoRepository<DepartmentEntity, String>{
+
+    Mono<DepartmentEntity> findById(@Param("id") String  id);
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/logic/ReactiveDepartmentCrud.java:10]: I don't know if this is necessary.
What happens if you delete this? does something break?

**Review by Noam-Levy — APPROVED** [review]

File: usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentBoundary.java [hunk 16]
```diff
@@ -44,21 +44,12 @@ public void setCreationDate(String creationDate) {
...
-
-        return outputDate;
-    }
 
     public DepartmentEntity toEntity() {
         DepartmentEntity entity = new DepartmentEntity();
         entity.setDeptId(this.getDeptId());
         entity.setDepartmentName(this.getDepartmentName());
-        entity.setCreationDate(changeDateFormat(LocalDate.now().toString()));
+        entity.setCreationDate(DateValidator.changeDateFormat(LocalDate.now().toString()));
```
Comment (Noam-Levy) [inline usersService/src/main/java/il/ac/afeka/usersservice/boundaries/DepartmentBoundary.java:52]: Why passing the date as a string only to parse it into date again?
you can pass it to the function as a date and getting it back as a string

**Pull request merged**